instrumentation/opentelemetry-annotations-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/otelannotations/async/Jdk8MethodStrategy.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.instrumentation.otelannotations.async;
+
+import io.opentelemetry.context.Context;
+import io.opentelemetry.instrumentation.api.tracer.BaseTracer;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+enum Jdk8MethodStrategy implements MethodSpanStrategy {
+  INSTANCE;
+
+  @Override
+  public boolean supports(Class<?> returnType) {
+    return returnType == CompletionStage.class || returnType == CompletableFuture.class;
+  }
+
+  @Override
+  public Object end(BaseTracer tracer, Context context, Class<?> returnType, Object result) {
+    if (result instanceof CompletableFuture) {
+      CompletableFuture<?> future = (CompletableFuture<?>) result;
+      if (future.isDone()) {
+        return endSynchronously(future, tracer, context);
+      }
```

Comment:
What's the reason for this block? I've tried removing it, added the following tests:

```groovy
  def "should capture span for completed CompletableFuture"() {
    setup:
    def future = CompletableFuture.completedFuture("Done")
    new TracedWithSpan().completableFuture(future)

    expect:
    assertTraces(1) {
      trace(0, 1) {
        span(0) {
          name "TracedWithSpan.completableFuture"
          kind SpanKind.INTERNAL
          hasNoParent()
          errored false
          attributes {
          }
        }
      }
    }
  }

  def "should capture span for completed exceptionally CompletableFuture"() {
    setup:
    def future = new CompletableFuture()
    future.completeExceptionally(new IllegalArgumentException("Boom"))
    new TracedWithSpan().completableFuture(future)

    expect:
    assertTraces(1) {
      trace(0, 1) {
        span(0) {
          name "TracedWithSpan.completableFuture"
          kind SpanKind.INTERNAL
          hasNoParent()
          errored true
          errorEvent(IllegalArgumentException, "Boom")
          attributes {
          }
        }
      }
    }
  }
```

and everything still passed.

Reply:
It's an optimization, one that can't really be detected in the tests.  It avoids the allocation of the callback lambda if the future is already completed.  It's not entirely necessary and I can remove it.

Reply:
Can you add a comment explaining why it was added? I think it's fine to leave it here, but it should be obvious why we're doing this.

Also, WDYT about adding tests that cover this path?

Reply:
Can do.